File: src/api/authTicket.ts
// Input: API client 与 ApiResponse
// Output: authTicketApi
// Pos: 授权票据 API 层
// 一旦我被更新，务必更新我的开头注释，以及所属的文件夹的 md。

import { client } from './client';
import { ApiResponse } from '../types';

/**
 * 授权范围
 */
export interface AuthScope {
    archiveYears?: number[];
    docTypes?: string[];
    keywords?: string[];
    accessType?: 'READ_ONLY' | 'READ_WRITE';
}

/**
 * 审批链
 */
export interface ApprovalChain {
    firstApprover?: {
        approverId: string;
        approverName: string;
        comment?: string;
        approved: boolean;
        approvedAt?: string;
    };
    secondApprover?: {
        approverId: string;
        approverName: string;
        comment?: string;
        approved: boolean;
        approvedAt?: string;
    };
}

/**
 * 授权票据详情
 */
export interface AuthTicketDetail {
    id: string;
    applicantId: string;
    applicantName: string;
    sourceFonds: string;
    targetFonds: string;
    scope: AuthScope;
    expiresAt: string; // ISO datetime string
    status: 'PENDING' | 'APPROVED' | 'REJECTED' | 'REVOKED' | 'EXPIRED';
    approvalChain?: ApprovalChain;
    reason: string;
    createdAt: string; // ISO datetime string
    lastModifiedTime?: string; // ISO datetime string
}

/**
 * 创建授权票据申请请求
 */
export interface CreateAuthTicketRequest {
    targetFonds: string;
    scope: AuthScope;
    expiresAt: string; // ISO datetime string
    reason: string;
}

/**
 * 审批请求
 */
export interface ApprovalRequest {
    comment: string;
    approved: boolean;
}

/**
 * 授权票据列表查询参数
 */
export interface AuthTicketListParams {
    page?: number;
    size?: number;
    status?: string;
    sourceFonds?: string;
    targetFonds?: string;
    applicantId?: string;
}

export const authTicketApi = {
    /**
     * 查询授权票据列表
     */
    list: async (params?: AuthTicketListParams): Promise<ApiResponse<{
        records: AuthTicketDetail[];
        total: number;
        page: number;
        size: number;
    }>> => {
        const response = await client.get<ApiResponse<{
            records: AuthTicketDetail[];
            total: number;
            page: number;
            size: number;
        }>>(
            '/auth-ticket/list',
            { params }
        );
        return response.data;
    },

    /**
     * 创建授权票据申请
     */
    apply: async (request: CreateAuthTicketRequest) => {
        const response = await client.post<ApiResponse<{
            ticketId: string;
            status: string;
            createdAt: string;
        }>>(
            '/auth-ticket/apply',
            request.scope,
            {
                params: {
                    targetFonds: request.targetFonds,
                    expiresAt: request.expiresAt,
                    reason: request.reason,
                },
            }
        );
        return response.data;
    },

    /**
     * 查询授权票据详情
     */
    getDetail: async (ticketId: string) => {
        const response = await client.get<ApiResponse<AuthTicketDetail>>(
            `/auth-ticket/${ticketId}`
        );
        return response.data;
    },

    /**
     * 撤销授权票据
     */
    revoke: async (ticketId: string, reason: string) => {
        const response = await client.post<ApiResponse<void>>(
            `/auth-ticket/${ticketId}/revoke?reason=${encodeURIComponent(reason)}`,
            null
        );
        return response.data;
    },

    /**
     * 第一审批人审批
     */
    firstApproval: async (ticketId: string, request: ApprovalRequest) => {
        const params = new URLSearchParams();
        params.append('comment', request.comment);
        params.append('approved', request.approved.toString());
        const response = await client.post<ApiResponse<void>>(
            `/auth-ticket/${ticketId}/first-approval?${params.toString()}`,
            null
        );
        return response.data;
    },

    /**
     * 第二审批人审批（复核）
     */
    secondApproval: async (ticketId: string, request: ApprovalRequest) => {
        const params = new URLSearchParams();
        params.append('comment', request.comment);
        params.append('approved', request.approved.toString());
        const response = await client.post<ApiResponse<void>>(
            `/auth-ticket/${ticketId}/second-approval?${params.toString()}`,
            null
        );
        return response.data;
    },
};

File: nexusarchive-java/src/main/java/com/nexusarchive/entity/AuthTicket.java
// Input: MyBatis-Plus、Lombok、Java 标准库
// Output: AuthTicket 类
// Pos: 领域实体/模型
// 一旦我被更新，务必更新我的开头注释，以及所属的文件夹的 md。

package com.nexusarchive.entity;

import com.baomidou.mybatisplus.annotation.*;
import lombok.Data;

import java.time.LocalDateTime;

/**
 * 跨全宗访问授权票据实体
 * 对应表: auth_ticket
 * 
 * PRD 来源: Section 2.4 - 跨全宗访问授权票据
 */
@Data
@TableName("auth_ticket")
public class AuthTicket {

    @TableId(type = IdType.ASSIGN_ID)
    private String id;

    /**
     * 申请人ID
     */
    private String applicantId;

    /**
     * 申请人姓名
     */
    private String applicantName;

    /**
     * 源全宗号（申请人所属全宗）
     */
    private String sourceFonds;

    /**
     * 目标全宗号
     */
    private String targetFonds;

    /**
     * 访问范围（JSON格式）
     * 格式: { "archiveYears": [2020, 2021], "docTypes": ["凭证"], "keywords": [], "accessType": "READ_ONLY" }
     */
    private String scope;

    /**
     * 有效期
     */
    private LocalDateTime expiresAt;

    /**
     * 状态: PENDING(待审批), FIRST_APPROVED(第一审批通过), APPROVED(已批准), 
     * REJECTED(已拒绝), REVOKED(已撤销), EXPIRED(已过期)
     */
    private String status;

    /**
     * 审批链快照（JSON格式）
     */
    private String approvalSnapshot;

    /**
     * 申请原因
     */
    private String reason;

    /**
     * 创建时间
     */
    @TableField(value = "created_at", fill = FieldFill.INSERT)
    private LocalDateTime createdAt;

    /**
     * 最后修改时间
     */
    @TableField(value = "last_modified_time", fill = FieldFill.INSERT_UPDATE)
    private LocalDateTime lastModifiedTime;

    /**
     * 逻辑删除标记
     */
    @TableLogic
    private Integer deleted;
}



